### DISCUSSION TITLE
Read FCStd node selected solid name

### DISCUSSION THREAD

**Discussion by macdroid53** [nortikin/sverchok#4459] [Ideas]
Is it possible to have the name of the selected model output and passed so that the the name is used instead of Alpha/Beta/etc. in the subsequent nodes and final mesh in the tree?

Possibly a new feature?

**Comment by Durman** [comment]
Do you mean something like this?

![image](https://user-images.githubusercontent.com/28003269/167297628-39b802c2-39a6-4f39-aa6a-6d692a2f9b2d.png)

**Reply by macdroid53** [reply]
Hmm...is this what that would do?
![image](https://user-images.githubusercontent.com/6831752/167301175-5c823a5e-2f6a-4a33-9062-caadc6bdf8e4.png)

**Comment by macdroid53** [comment]
Is this a crazy idea? It seems this thread has gone quiet...

**Reply by Durman** [reply]
I just did not use Free Cad nodes. Probably other developers can comment on that.

**Reply by zeffii** [reply]
we could subclass MeshViewer and make it accept Solids.
unfortunately the SvSolidsSocket current passes `obj.Shape` , and that doesn't carry name information (From what i can tell)

![image](https://user-images.githubusercontent.com/619340/170991039-58c8b479-286b-403f-8c8c-fbeb6e8d710e.png)

The request would take a small rewrite @macdroid53 , if someone finds time / feels like doing it  it can probably be done without interfering with existing blend files.

**Reply by macdroid53** [reply]
I was thinking more of a name output on the Read FCStd node and a name input on the Mesh viewer...since the user needs to select the name of the solid that is passed to the Solid to Mesh node. But, I'm sure I don't see the bigger picture...

**Reply by zeffii** [reply]
> name output on the Read FCStd node and a name input on the Mesh viewer.

that would solve the immediate problem, no doubt.

**Reply by zeffii** [reply]
Instead of passing around `[obj.Shape, ...]` it might be more useful to pass around `[obj, ...]` , or some kind of wrapper object which does include the `Label` attribute of the object - i'm a little surprised there's no way to the get `Label` from `obj.Shape`.

below `obj` is a `solid` like:
```python
                    for obj in F.ActiveDocument.Objects:
                        if obj.Module in obj_mask or obj.TypeId in obj_mask:
                            ... = obj 
```
```python
from dataclasses import dataclass

@dataclass
class SolidObjectStreamItem:
    shape: object
    label: str


wrapped_obj = SolidObjectStreamItem(obj.Shape, obj.Label)
wrapped_obj.shape
>>> <Solid object at 0,000,020,39F,A23,EC0>

wrapped_obj.label
>>> "Base"
```
using the lowercase attribute names allows us to make it clear that these from the wrapper and not the Solid's API

**Comment by zeffii** [comment]
there's one way to already do this, if you like "temporary solutions (tm)".

```python
"""
in solid_in So
in obj_in o
"""

my_input = self.inputs['solid_in']
if my_input.is_linked:
    origin_node = my_input.other.node

    if (label := getattr(origin_node, "selected_label", None)):
        print(label)
    
    if (my_obj_writer := self.inputs['obj_in']).is_linked:
        writer_node = my_obj_writer.other.node
        writer_node.base_data_name = label
```
![image](https://user-images.githubusercontent.com/619340/171407399-7ac2a90c-d08f-4291-975e-d03cc9bf90f8.png)

**Reply by zeffii** [reply]
or 
```python
"""
in solid_in So
in obj_in o
"""

if (my_input := self.inputs['solid_in']).is_linked:
    origin_node = my_input.other.node

    label = getattr(origin_node, "selected_label", None)
    
    if label and (my_obj_writer := self.inputs['obj_in']).is_linked:
        writer_node = my_obj_writer.other.node
        writer_node.base_data_name = label
```
